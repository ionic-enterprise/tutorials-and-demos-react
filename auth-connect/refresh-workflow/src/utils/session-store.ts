import { AuthResult } from '@ionic-enterprise/auth';
import { Preferences } from '@capacitor/preferences';

let session: AuthResult | null = null;
let listeners: (() => void)[] = [];

const subscribe = (listener: () => void) => {
  listeners = [...listeners, listener];
  return () => {
    listeners = listeners.filter((l) => l !== listener);
  };
};

const getSnapshot = (): AuthResult | null => {
  return session;
};

const emitChange = () => {
  for (const listener of listeners) {
    listener();
  }
};

const setSession = async (newSession: AuthResult | null) => {
  session = newSession;
  await Preferences.set({ key: 'session', value: JSON.stringify(newSession) });
  emitChange();
};

const initialize = async (): Promise<void> => {
  const { value } = await Preferences.get({ key: 'session' });
  if (value) {
    session = JSON.parse(value);
  }
};

export { initialize, subscribe, getSnapshot, setSession };
