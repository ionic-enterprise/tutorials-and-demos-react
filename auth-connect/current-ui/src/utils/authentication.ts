import { Capacitor } from '@capacitor/core';
import { Auth0Provider, AuthConnect, AzureProvider, ProviderOptions } from '@ionic-enterprise/auth';
import { getSnapshot, setSession } from './session-store';

const isNative = Capacitor.isNativePlatform();
// const provider = new Auth0Provider();

// const authOptions: ProviderOptions = {
//   audience: 'https://io.ionic.demo.ac',
//   clientId: 'yLasZNUGkZ19DGEjTmAITBfGXzqbvd00',
//   discoveryUrl: 'https://dev-2uspt-sz.us.auth0.com/.well-known/openid-configuration',
//   logoutUrl: isNative ? 'io.ionic.acdemo://auth-action-complete' : 'http://localhost:8100/auth-action-complete',
//   redirectUri: isNative ? 'io.ionic.acdemo://auth-action-complete' : 'http://localhost:8100/auth-action-complete',
//   scope: 'openid offline_access email picture profile',
// };

const provider = new AzureProvider();
const authOptions: ProviderOptions = {
  clientId: 'ed8cb65d-7bb2-4107-bc36-557fb680b994',
  scope:
    'openid offline_access email profile https://dtjacdemo.onmicrosoft.com/ed8cb65d-7bb2-4107-bc36-557fb680b994/demo.read',
  discoveryUrl:
    'https://dtjacdemo.b2clogin.com/dtjacdemo.onmicrosoft.com/v2.0/.well-known/openid-configuration?p=B2C_1_acdemo2',
  redirectUri: isNative
    ? 'msauth://com.ionic.acprovider/O5m5Gtd2Xt8UNkW3wk7DWyKGfv8%3D'
    : 'http://localhost:8100/auth-action-complete',
  logoutUrl: isNative
    ? 'msauth://com.ionic.acprovider/O5m5Gtd2Xt8UNkW3wk7DWyKGfv8%3D'
    : 'http://localhost:8100/auth-action-complete',
  audience: '',
};

const setupAuthConnect = async (): Promise<void> => {
  return AuthConnect.setup({
    platform: isNative ? 'capacitor' : 'web',
    logLevel: 'DEBUG',
    ios: { webView: 'private' },
    web: { uiMode: 'current', authFlow: 'PKCE' },
  });
};

const handleAuthCallback = async (): Promise<void> => {
  const params = new URLSearchParams(window.location.search);
  if (params.size > 0) {
    const queryEntries = Object.fromEntries(params.entries());
    const authResult = await AuthConnect.handleLoginCallback(queryEntries, authOptions);
    setSession(authResult);
  } else {
    setSession(null);
  }
};

const login = async (): Promise<void> => {
  const authResult = await AuthConnect.login(provider, authOptions);
  setSession(authResult);
};

const logout = async (): Promise<void> => {
  const authResult = getSnapshot();
  if (authResult) {
    await AuthConnect.logout(provider, authResult);
    setSession(null);
  }
};

export { handleAuthCallback, login, logout, setupAuthConnect };
