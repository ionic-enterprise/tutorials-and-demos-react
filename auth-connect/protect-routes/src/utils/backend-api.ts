import axios, { InternalAxiosRequestConfig } from 'axios';
import { getSnapshot, setSession } from './session-store';

const baseURL = 'https://cs-demo-api.herokuapp.com';

const client = axios.create({
  baseURL,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

client.interceptors.request.use(async (config: InternalAxiosRequestConfig) => {
  const session = getSnapshot();
  if (session?.accessToken && config.headers) {
    config.headers.Authorization = `Bearer ${session.accessToken}`;
  }
  return config;
});

client.interceptors.response.use(
  (response) => response,
  (error) => {
    if (error.response.status === 401) {
      setSession(null);
    }
    return Promise.reject(error);
  },
);

export { client };
