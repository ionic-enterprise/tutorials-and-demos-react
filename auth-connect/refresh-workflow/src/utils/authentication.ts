import { Capacitor } from '@capacitor/core';
import { Auth0Provider, AuthConnect, AuthResult, ProviderOptions } from '@ionic-enterprise/auth';
import { getSnapshot, setSession } from './session-store';

const isNative = Capacitor.isNativePlatform();
const provider = new Auth0Provider();

const authOptions: ProviderOptions = {
  audience: 'https://io.ionic.demo.ac',
  clientId: 'yLasZNUGkZ19DGEjTmAITBfGXzqbvd00',
  discoveryUrl: 'https://dev-2uspt-sz.us.auth0.com/.well-known/openid-configuration',
  logoutUrl: isNative ? 'io.ionic.acdemo://auth-action-complete' : 'http://localhost:8100/auth-action-complete',
  redirectUri: isNative ? 'io.ionic.acdemo://auth-action-complete' : 'http://localhost:8100/auth-action-complete',
  scope: 'openid offline_access email picture profile',
};

const setupAuthConnect = async (): Promise<void> => {
  return AuthConnect.setup({
    platform: isNative ? 'capacitor' : 'web',
    logLevel: 'DEBUG',
    ios: { webView: 'private' },
    web: { uiMode: 'popup', authFlow: 'PKCE' },
  });
};

const login = async (): Promise<void> => {
  const authResult = await AuthConnect.login(provider, authOptions);
  setSession(authResult);
};

const logout = async (): Promise<void> => {
  const authResult = getSnapshot();
  if (authResult) {
    await AuthConnect.logout(provider, authResult);
    setSession(null);
  }
};

const refreshIfExpired = async (): Promise<void> => {
  const authResult = getSnapshot();
  console.log('check the refresh', authResult);
  if (
    authResult &&
    (await AuthConnect.isAccessTokenAvailable(authResult)) &&
    (await AuthConnect.isAccessTokenExpired(authResult))
  ) {
    let newAuthResult: AuthResult | null = null;
    if (await AuthConnect.isRefreshTokenAvailable(authResult)) {
      try {
        newAuthResult = await AuthConnect.refreshSession(provider, authResult);
      } catch (err) {
        console.error(err);
      }
    }
    setSession(newAuthResult);
  }
};

export { login, logout, refreshIfExpired, setupAuthConnect };
