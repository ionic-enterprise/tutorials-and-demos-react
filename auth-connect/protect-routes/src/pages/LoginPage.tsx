import { IonButton, IonContent, IonHeader, IonPage, IonTitle, IonToolbar } from '@ionic/react';
import { login } from '../utils/authentication';
import { useHistory } from 'react-router';

const LoginPage: React.FC = () => {
  const history = useHistory();
  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonTitle>Login</IonTitle>
        </IonToolbar>
      </IonHeader>
      <IonContent fullscreen>
        <IonHeader collapse="condense">
          <IonToolbar>
            <IonTitle size="large">Login</IonTitle>
          </IonToolbar>
        </IonHeader>
        <IonButton
          onClick={async () => {
            await login();
            history.push('tabs/tab1');
          }}
        >
          Login
        </IonButton>
      </IonContent>
    </IonPage>
  );
};

export default LoginPage;
