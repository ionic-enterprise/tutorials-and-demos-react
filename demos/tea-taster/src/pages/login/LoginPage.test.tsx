import { Mock, vi } from 'vitest';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import LoginPage from './LoginPage';
import { useHistory } from 'react-router-dom';
import { login } from '../../utils/auth';

vi.mock('react-router-dom');
vi.mock('../../utils/auth');

describe('<LoginPage />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders consistently', () => {
    const { asFragment } = render(<LoginPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the title', () => {
    render(<LoginPage />);
    const titleElements = screen.getAllByText('Login');
    expect(titleElements).toHaveLength(1);
  });

  describe('sign in button', () => {
    it('starts disabled', async () => {
      render(<LoginPage />);
      const button = await waitFor(() => screen.getByText('Sign In') as HTMLIonButtonElement);
      await waitFor(() => expect(button.disabled).toBeTruthy());
    });

    it('is enabled once valid data is entered', async () => {
      render(<LoginPage />);
      const button = await waitFor(() => screen.getByText('Sign In') as HTMLIonButtonElement);
      const password = await waitFor(() => screen.getByLabelText('Password'));
      const email = await waitFor(() => screen.getByLabelText('Email Address'));
      await waitFor(() => expect(button.disabled).toBeTruthy());
      await waitFor(() => fireEvent.input(email, { target: { value: 'test@test.com' } }));
      await waitFor(() => expect(button.disabled).toBeTruthy());
      await waitFor(() => fireEvent.input(password, { target: { value: 'password' } }));
      await waitFor(() => expect(button.disabled).toBeFalsy());
    });

    describe('clicking the sign in button', () => {
      let toast: HTMLIonToastElement;

      beforeEach(async () => {
        render(<LoginPage />);
        const email = await waitFor(() => screen.getByLabelText('Email Address'));
        const password = await waitFor(() => screen.getByLabelText('Password'));
        await waitFor(() => fireEvent.input(email, { target: { value: 'test@test.com' } }));
        await waitFor(() => fireEvent.input(password, { target: { value: 'password' } }));
        toast = await waitFor(() => screen.getByTestId('error-toast') as HTMLIonToastElement);
      });

      it('performs the login', async () => {
        const button = await waitFor(() => screen.getByText('Sign In') as HTMLIonButtonElement);
        fireEvent.click(button);
        await waitFor(() => expect(login).toHaveBeenCalledTimes(1));
        expect(login).toHaveBeenCalledWith('test@test.com', 'password');
      });

      describe('if the login succeeds', () => {
        beforeEach(() => (login as Mock).mockResolvedValue(true));

        it('does not show an error', async () => {
          const button = await waitFor(() => screen.getByText('Sign In') as HTMLIonButtonElement);
          fireEvent.click(button);
          await waitFor(() => expect(Array.from(toast.classList)).toContain('overlay-hidden'));
        });

        it('navigates to the root page', async () => {
          const history = useHistory();
          const button = await waitFor(() => screen.getByText('Sign In') as HTMLIonButtonElement);
          fireEvent.click(button);
          await waitFor(() => expect(history.replace).toBeCalledTimes(1));
          await waitFor(() => expect(history.replace).toHaveBeenCalledWith('/'));
        });
      });

      describe('if the login fails', () => {
        beforeEach(() => (login as Mock).mockResolvedValue(false));

        it('shows an error', async () => {
          const button = await waitFor(() => screen.getByText('Sign In') as HTMLIonButtonElement);
          fireEvent.click(button);
          await waitFor(() => expect(Array.from(toast.classList)).not.toContain('overlay-hidden'));
        });

        it('does not navigate', async () => {
          const history = useHistory();
          const button = await waitFor(() => screen.getByText('Sign In') as HTMLIonButtonElement);
          fireEvent.click(button);
          await waitFor(() => expect(history.replace).not.toHaveBeenCalled());
        });
      });
    });
  });

  describe('error messages', () => {
    it('displays an error if the e-mail address is dirty and empty', async () => {
      render(<LoginPage />);
      const email = await waitFor(() => screen.getByLabelText('Email Address'));
      await waitFor(() => fireEvent.input(email, { target: { value: 'test@test.com' } }));
      await waitFor(() => fireEvent.blur(email));
      await waitFor(() => fireEvent.input(email, { target: { value: '' } }));
      await waitFor(() => expect(screen.getByText(/Email address is a required field/)).toBeInTheDocument());
    });

    it('displays an error if the email address has an invalid format', async () => {
      render(<LoginPage />);
      const email = await waitFor(() => screen.getByLabelText('Email Address'));
      await waitFor(() => fireEvent.input(email, { target: { value: 'test' } }));
      await waitFor(() => fireEvent.blur(email));
      await waitFor(() => expect(screen.getByText(/Email address must be a valid email/)).toBeInTheDocument());
    });

    it('displays an error message if the password is dirty and empty', async () => {
      render(<LoginPage />);
      const password = await waitFor(() => screen.getByLabelText('Password'));
      await waitFor(() => fireEvent.input(password, { target: { value: '' } }));
      await waitFor(() => fireEvent.blur(password));
      await waitFor(() => expect(screen.getByText(/Password is a required field/)).toBeInTheDocument());
    });
  });
});
