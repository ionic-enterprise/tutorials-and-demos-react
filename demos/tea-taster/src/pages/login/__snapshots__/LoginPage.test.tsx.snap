// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<LoginPage /> > renders consistently 1`] = `
<DocumentFragment>
  <div
    class="ion-page"
  >
    <ion-content
      class="login-page main-content"
      fullscreen="true"
    >
      <form>
        <ion-card>
          <ion-card-header>
            <ion-card-title>
              Login
            </ion-card-title>
            <ion-card-subtitle>
              Welcome to Tea Taster
            </ion-card-subtitle>
          </ion-card-header>
          <ion-card-content>
            <ion-input
              class="ion-valid ion-untouched ion-pristine"
              label="Email Address"
              label-placement="floating"
              type="email"
            >
              <!---->
            </ion-input>
            <ion-input
              class="ion-valid ion-untouched ion-pristine"
              label="Password"
              label-placement="floating"
              type="password"
            >
              <!---->
            </ion-input>
            <ion-button
              disabled="true"
              expand="full"
            >
              Sign In
              <ion-icon
                icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path d='M192 176v-40a40 40 0 0140-40h160a40 40 0 0140 40v240a40 40 0 01-40 40H240c-22.09 0-48-17.91-48-40v-40' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/><path stroke-linecap='round' stroke-linejoin='round' d='M288 336l80-80-80-80M80 256h272' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
                slot="end"
              />
            </ion-button>
          </ion-card-content>
        </ion-card>
      </form>
      <template />
    </ion-content>
  </div>
</DocumentFragment>
`;
